hack/test-go.sh

```diff
@@ -1,125 +1,231 @@
+              -o -wholename '*_tools' \
+              -o -wholename '*.git' \
+              -o -wholename '*openshift.local.*' \
+              -o -wholename '*Godeps/*' \
+              -o -wholename '*assets/node_modules' \
+              -o -wholename '*test/extended' \
+        \) -prune \
+    \) -name '*_test.go' -print0 | xargs -0n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n"
+}
 
-    go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "$KUBE_COVER_PROFILE" "$test_package" "${@:2}"
-  done
+# Determine packages to test
+test_packages=
+if [[ "$#" -gt 0 ]]; then
+    for package in "${@:1}"; do
+        if [[ "${package}" = *"/..." ]]; then
+            # this is the 'go test' syntax for "recursively test all under this path" but since we symlink
+            # our source files to the _output directory, 'go test' can't expand this syntax correctly, so
+            # we do it ourselves
+            subpackages="$( list_test_packages_under "${package%"/..."}" )"
+            for subpackage in "${subpackages}"; do
+                test_packages="${test_packages} ${subpackage}"
+            done
+        else
+            test_packages="${test_packages} ${OS_GO_PACKAGE}/${package}"
+        fi
+    done
+else
+    # If no packages are given to test, we need to generate a list of all packages with unit tests
+    openshift_test_packages="$(list_test_packages_under '*')"  
 
-  echo 'mode: atomic' > ${OUTPUT_COVERAGE}/profiles.out
-  find $OUTPUT_COVERAGE -name profile.out | xargs sed '/^mode: atomic$/d' >> ${OUTPUT_COVERAGE}/profiles.out
-  go tool cover "-html=${OUTPUT_COVERAGE}/profiles.out" -o "${OUTPUT_COVERAGE}/coverage.html"
+    kubernetes_path="Godeps/_workspace/src/k8s.io/kubernetes"
+    mandatory_kubernetes_packages="$( echo "${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api"; echo "${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api/v1"; )"
 
-  rm -rf $OUTPUT_COVERAGE/$OS_GO_PACKAGE
-else
-  nice go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "${@:2}" $test_packages
+    test_packages="$(echo "${openshift_test_packages}"; echo "${mandatory_kubernetes_packages}")"
+
+    if [[ -n "${test_kube}" ]]; then
+        optional_kubernetes_packages="$(find "${kubernetes_path}" -not \( \
+          \( \
+            -wholename "${kubernetes_path}/pkg/api" \
+            -o -wholename "${kubernetes_path}/pkg/api/v1" \
+            -o -wholename "${kubernetes_path}/test/e2e" \
+          \) -prune \
+        \) -name '*_test.go' -print0 | xargs -0n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n")"
+
+        test_packages="$(echo "${test_packages}"; echo "${optional_kubernetes_packages}")"
+    fi
 fi
 
-ret=$?; ENDTIME=$(date +%s); echo "$0 took $(($ENDTIME - $STARTTIME)) seconds"; exit "$ret"
+if [[ -n "${print_packages}" ]]; then
+    echo "The following packages will be tested by $0:"
+    for package in "${test_packages}"; do
+        echo "${package}"
+    done
+    exit 0
```

Comment:
Ah, so `PRINT_PACKAGES` is kind of a `--dry-run` flag.  I see now that that is explained above.  I kind of wish the name of the parameter or the output better indicated that, but you can argue that I should have just paid more attention to the description of the parameter at the top of the file, so I'm ambivalent.